rust/src/dataset/write.rs

```diff
@@ -53,3 +73,300 @@ impl Default for WriteParams {
         }
     }
 }
+
+/// Convert reader to a stream
+///
+/// Will peek the first batch to get the dictionaries for dictionary columns
+pub fn reader_to_stream(
```

Comment:
this is a `crate` function? 

Also how could we make sure that the schema constructed from arrow::Schema is the same as dataset schema (i.e., there might be holes w.r..t of field ids).

In other word, do we expect this function is only be called by `CREATE`?

Reply:
> this is a crate function?

Right. This module is private, but in the super module we reexport `*`. I can change the re-export to be selective. (Using `pub(crate)` in a private module triggers a lint.)

> Also how could we make sure that the schema constructed from arrow::Schema is the same as dataset schema (i.e., there might be holes w.r..t of field ids).

Hmm I see what you mean, but I also did not change any logic nor the tests. This seems to be a pre-existing issue, no? I can look to see if there is an easy fix.

Reply:
It looks like how this is handled is here:

https://github.com/lancedb/lance/blob/b663c2caef9129d80537ba7dc2b8b9138173e38e/rust/src/dataset.rs#L324-L335

So in the append case, the caller needs to verify the resulting schema matches the existing table schema. Maybe it's sufficient to call that out in the documentation of this function?